nim_status/lib/permissions.nim

```diff
@@ -0,0 +1,92 @@
+
+import # vendor libs
+  sqlcipher, json_serialization, stew/byteutils
+from stew/shims/macros as stew_macros import hasCustomPragmaFixed, getCustomPragmaFixed
+
+import # nim-status libs
+  ../types, conversions
+
+type Dapp* {.dbTableName("dapps").} = object
+  name* {.dbColumnName("name").}: string
+
+type DappPermissions* {.dbTableName("permissions").} = object
+  name* {.
+    serializedFieldName("dapp"),
+    dbColumnName("dapp_name"),
+    dbForeignKey(Dapp)
+  .}: string
+  permissions* {.
+    serializedFieldName("permissions"),
+    dbColumnName("permission")
+  .}: seq[string]
+
+
+proc addPermissions*(db: DbConn, dappPerms: DappPermissions) =
+  var dapp: Dapp
+  var dappPermission: DappPermissions
+  var query = fmt"""INSERT OR REPLACE INTO {dapp.tableName}
+                      (
+                        {dapp.name.columnName}
+                      )
+                    VALUES (?)"""
+
+  db.exec(query, dappPerms.name)
+
+  if dappPerms.permissions.len == 0:
+    return
+
+  query = fmt"""INSERT INTO {dappPermission.tableName}
+                  (
+                    {dappPermission.name.columnName},
+                    {dappPermission.permissions.columnName}
+                  )
+                VALUES(?, ?)"""
+  for perm in dappPerms.permissions:
+    db.exec(query, dappPerms.name, @[perm])
+
+proc getPermissions*(db: DbConn): seq[DappPermissions] =
+  var
+    dapp: Dapp
+    dappPermission: DappPermissions
+  var query = fmt"""SELECT {dapp.name.columnName}
+                    FROM {dapp.tableName}"""
+
+  let dapps = execQuery[Dapp](db, query)
+  if dapps.len == 0:
+    return result
+
+  # let tblDappPerms = initTable[string, DappPermissions]()
+  var tblDappPerms = dapps.map(dapp => (dapp.name, DappPermissions(name: dapp.name))).toOrderedTable()
+  # for dapp in dapps:
+  #   tb
+
+  query = fmt"""SELECT 
+                      {dappPermission.name.columnName},
+                      {dappPermission.permissions.columnName}
+                    FROM {dappPermission.tableName}"""
+
+  let dappPerms = execQuery[DappPermissions](db, query)
+  for dappPerm in dappPerms:
+    if not tblDappPerms.hasKey(dappPerm.name):
+      continue
+    var existing = tblDappPerms[dappPerm.name]
+    ## existing.permissions = existing.permissions & dappPerm.permissions
+    # existing.permissions.insert(dappPerm.permissions, existing.permissions.len)
+    existing.permissions = existing.permissions & dappPerm.permissions
+    tblDappPerms[dappPerm.name] = existing
+  # dappPerms.map(dappPerm => tblDappPerms[dappPerm.name].insert dappPerm.permissions, tblDappPerms[dapp])
+  toSeq(tblDappPerms.values)
+
+proc deletePermission*(db: DbConn, name: string) =
+  var dapp: Dapp
+  let query = fmt"""DELETE FROM
+                      {dapp.tableName}
+                    WHERE
+                      {dapp.name.columnName} = ?"""
+
+  db.exec(query, name)
+  # TODO: Should we also delete permissions from the permissions table?
```

Comment:
Probably not necessary due to the table having `FOREIGN KEY(dapp_name) REFERENCES dapps(name) ON DELETE CASCADE`